app/services/github/client.rb

```diff
@@ -0,0 +1,98 @@
+          end
+        end
+
+        # call the original method, this will only be called the first time
+        # as in subsequent calls, the newly defined method will prevail
+        request do
+          target.public_send(method, *args, &block)
+        end
+      end
+
+      # adapted from https://api.rubyonrails.org/classes/Module.html#method-i-delegate_missing_to
+      def respond_to_missing?(method, _include_all = false)
+        target.respond_to?(method, false) || super
+      end
+
+      private
+
+      def request
+        Honeycomb.add_field("app.name", "github.client")
+        yield
+      rescue Octokit::Error => e
+        class_name = e.class.name.demodulize
+
+        record_error(e.message, class_name)
+
+        # raise specific error if known, generic one if unknown
+        error_class = "::Github::Errors::#{class_name}".safe_constantize
+        raise error_class, e.message if error_class
+
+        error_class = if e.class < Octokit::ClientError
+                        Github::Errors::ClientError
+                      elsif e.class < Octokit::ServerError
+                        Github::Errors::ServerError
+                      else
+                        Github::Errors::Error
+                      end
+
+        raise error_class, e.message
+      end
+
+      def record_error(error_message, class_name)
+        Honeycomb.add_field("github.result", "error")
+        Honeycomb.add_field("github.error", class_name)
+        DatadogStatsClient.increment(
+          "github.errors",
+          tags: ["error:#{class_name}", "message:#{error_message}"],
+        )
+      end
+
+      def target
+        @target ||= Octokit::Client.new(
+          client_id: ApplicationConfig["GITHUB_KEY"],
+          client_secret: ApplicationConfig["GITHUB_SECRET"],
+          user_agent: "#{Octokit::Default::USER_AGENT} (#{URL.url})",
+          middleware: faraday_middleware_stack,
+          connection_options: connection_options,
+        )
+      end
+
+      def faraday_middleware_stack
+        # Extending the default functionality
+        # see <https://github.com/octokit/octokit.rb#advanced-usage>,
+        # <https://github.com/octokit/octokit.rb#caching>
+        # and <https://github.com/octokit/octokit.rb/blob/master/lib/octokit/default.rb>
+        Faraday::RackBuilder.new do |builder|
+          builder.use Faraday::HttpCache, store: Rails.cache, serializer: Marshal, shared_cache: false
+          builder.use Faraday::Request::Retry, exceptions: [Octokit::ServerError]
+          builder.use Octokit::Middleware::FollowRedirects
+          builder.use Octokit::Response::RaiseError
+          builder.use Octokit::Response::FeedParser
+
+          builder.adapter :patron
+        end
+      end
+
+      def connection_options
+        Octokit.connection_options.merge(
+          request: {
+            open_timeout: 5,
+            timeout: 5
+          },
```

Comment:
there's no timeout by default, there are strong reasons to have default on HTTP clients as we aren't in control of the upstream server anyway (GH can go down or can take long to respond). Event Octokit [recommends adding them](https://github.com/octokit/octokit.rb#timeouts):

> By default, Octokit does not timeout network requests.
> You should set a timeout in order to avoid Ruby’s Timeout module, which can hose your server.